src/shogun/preprocessor/RandomFourierGaussPreproc.cpp

```diff
@@ -8,393 +8,131 @@
-				"float64_t * CRandomFourierGaussPreproc::apply_to_feature_vector(...): test_rfinited()==false: you need to call before CRandomFourierGaussPreproc::init (Features *f) OR	1. set_dim_feature_space(const int32 dim), 2. set_dim_input_space(const int32 dim), 3. init_randomcoefficients() or set_randomcoefficients(...) \n");
-	}
+	SGVector<float64_t> res(m_dims);
+	SGVector<float64_t> offset_vector(m_dims);
 
-	float64_t val = std::sqrt(2.0 / cur_dim_feature_space);
-	SGVector<float64_t> res(cur_dim_feature_space);
-
-	for (int32_t od = 0; od < cur_dim_feature_space; ++od) {
-		SGVector<float64_t> wrapper(randomcoeff_multiplicative+od*cur_dim_input_space, cur_dim_input_space, false);
-		res[od] = val * cos(randomcoeff_additive[od] + linalg::dot(vector, wrapper));
-	}
+	assert_fitted();
+	linalg::add(m_offset,linalg::matrix_prod(m_basis,vector),offset_vector);
+	res=linalg::scale(linalg::cos(offset_vector),std::sqrt(2.0 / m_dims));
 
 	return res;
 }
 
-SGMatrix<float64_t>
-RandomFourierGaussPreproc::apply_to_matrix(SGMatrix<float64_t> matrix)
+SGMatrix<float64_t> RandomFourierGaussPreproc::apply_to_matrix(SGMatrix<float64_t> matrix)
 {
-	// version for case dim_feature_space < dim_input space with direct transformation on feature matrix ??
-
 	auto num_vectors = matrix.num_cols;
 	auto num_features = matrix.num_rows;
 
-	io::info("get Feature matrix: {}x{}", num_vectors, num_features);
+	assert_fitted();
+	SG_DEBUG("get Feature matrix: {}x{}", num_vectors, num_features);
 
-	if (num_features != cur_dim_input_space)
-	{
-		throw ShogunException(
-		    "float64_t * "
-		    "CRandomFourierGaussPreproc::apply_to_matrix("
-		    "SGMatrix<float64_t> matrix): matrix.num_rows != "
-		    "cur_dim_input_space is not allowed\n");
-	}
+	SGMatrix<float64_t> res(m_dims, num_vectors);
 
-	SGMatrix<float64_t> res(cur_dim_feature_space, num_vectors);
+	/* Temporary matrix to store offsetted matrix product*/
+	SGMatrix<float64_t> offset_mat(m_dims, num_vectors);
 
-	auto val = std::sqrt(2.0 / cur_dim_feature_space);
+	linalg::add_vector(
+	    linalg::matrix_prod(m_basis, matrix), m_offset, offset_mat);
+	res = linalg::scale(linalg::cos(offset_mat), std::sqrt(2.0 / m_dims));
 
-	for (auto vec : range(num_vectors))
-	{
-		for (auto od : range(cur_dim_feature_space))
+	return res;	
+}
+
+void RandomFourierGaussPreproc::sample_spectral_density(int32_t dim_input_space)
+{
+	NormalDistribution<float64_t> normal_dist;
+	UniformRealDistribution<float64_t> uniform_real_dist(0.0, 2 * Math::PI);
+	
+	for(auto row : range(m_dims))
+	{	
+		for(auto col : range(dim_input_space))
 		{
-			SGVector<float64_t> a(
-			    matrix.matrix + vec * num_features, cur_dim_input_space, false);
-			SGVector<float64_t> b(
-			    randomcoeff_multiplicative + od * cur_dim_input_space,
-			    cur_dim_input_space, false);
-			res(od, vec) =
-			    val * cos(randomcoeff_additive[od] + linalg::dot(a, b));
+			m_basis(row,col)= std::sqrt((float64_t)1 / m_kernelwidth) *
```

Comment:
could you compute the coefficient before the loop into a variable?, makes it easier to read